fireant/slicer/operations.py

```diff
@@ -1,85 +1,67 @@
-# coding: utf8
+import numpy as np
+import pandas as pd
 
+from .metrics import Metric
 
-class Operation(object):
-    """
-    The `Operation` class represents an operation in the `Slicer` API.
-    """
-    key = None
-    label = None
-
-    def schemas(self):
-        pass
 
-    def metrics(self):
-        return []
+def _extract_key_or_arg(data_frame, key):
+    return data_frame[key] \
+        if key in data_frame \
+        else key
 
 
-class Totals(Operation):
+class Operation(object):
     """
-    `Operation` for rolling up totals across dimensions in queries.  This will append the totals across a dimension to
-    a dimension.
+    The `Operation` class represents an operation in the `Slicer` API.
     """
-    key = '_total'
-    label = 'Total'
 
-    def __init__(self, *dimension_keys):
-        self.dimension_keys = dimension_keys
+    def apply(self, data_frame):
+        raise NotImplementedError()
 
 
-class L1Loss(Operation):
-    """
-    Performs L1 Loss (mean abs. error) operation on a metric using another metric as the target.
-    """
-    key = 'l1loss'
-    label = 'L1 loss'
+class _Cumulative(Operation):
+    def __init__(self, arg):
+        self.arg = arg
 
-    def __init__(self, metric_key, target_metric_key):
-        self.metric_key = metric_key
-        self.target_metric_key = target_metric_key
+    @property
+    def metrics(self):
+        return [metric
+                for metric in [self.arg]
+                if isinstance(metric, Metric)]
 
-    def schemas(self):
-        return {
-            'key': self.key,
-            'metric': self.metric_key,
-            'target': self.target_metric_key,
-        }
+    def apply(self, data_frame):
+        raise NotImplementedError()
 
-    def metrics(self):
-        return [self.metric_key, self.target_metric_key]
 
+class CumSum(_Cumulative):
+    def apply(self, data_frame):
+        if isinstance(data_frame.index, pd.MultiIndex):
+            return data_frame[self.arg.key] \
+                .groupby(level=data_frame.index.names[1:]) \
```

Comment:
Out of interest, what is the reason for slicing from the first element on in the index? Might be worth a comment.

Reply:
I'll write a comment. Basically, you don't want to continuing the cumulative function across different dimensions, you want every set of values for each dimension to start counting at zero. For example, given a device_type dimension, you would accumulate the values for mobile, then start at zero again for tablet.